src/Features/Core/Portable/Completion/CompletionService.ProviderManager.cs

```diff
@@ -57,41 +54,74 @@ public ProviderManager(CompletionService service)
-
-#pragma warning disable 0618
-                // We need to keep supporting built-in providers for a while longer since this is a public API.
-                // https://github.com/dotnet/roslyn/issues/42367
-                var builtin = _service.GetBuiltInProviders();
-#pragma warning restore 0618
+                if (project?.Solution.Workspace.Kind == WorkspaceKind.Interactive)
+                {
+                    // TODO (https://github.com/dotnet/roslyn/issues/4932): Don't restrict completions in Interactive
+                    return ImmutableArray<CompletionProvider>.Empty;
+                }
 
-                return imported.Concat(builtin).ToImmutableArray();
+                return ProjectCompletionProvider.GetExtensions(project);
             }
 
-            public ImmutableArray<CompletionProvider> GetProviders(ImmutableHashSet<string>? roles)
+            private ImmutableArray<CompletionProvider> GetImportedAndBuiltInProviders(ImmutableHashSet<string>? roles)
             {
                 roles ??= ImmutableHashSet<string>.Empty;
 
                 lock (_gate)
                 {
-                    return _rolesToProviders.GetOrAdd(roles, _createRoleProviders);
+                    if (!_rolesToProviders.TryGetValue(roles, out var providers))
+                    {
+                        providers = GetImportedAndBuiltInProvidersWorker(roles);
+                        _rolesToProviders.Add(roles, providers);
+                    }
+
+                    return providers;
+                }
+
+                ImmutableArray<CompletionProvider> GetImportedAndBuiltInProvidersWorker(ImmutableHashSet<string> roles)
+                {
+                    using var _ = ArrayBuilder<CompletionProvider>.GetInstance(out var providers);
+                    providers.AddRange(GetLazyImportedProviders()
+                        .Where(lz => lz.Metadata.Roles == null || lz.Metadata.Roles.Length == 0 || roles.Overlaps(lz.Metadata.Roles))
+                        .Select(lz => lz.Value));
+
+#pragma warning disable 0618
+                    // We need to keep supporting built-in providers for a while longer since this is a public API.
+                    // https://github.com/dotnet/roslyn/issues/42367
+                    providers.AddRange(_service.GetBuiltInProviders());
+#pragma warning restore 0618
+
+                    return providers.ToImmutable();
                 }
             }
 
-            public CompletionProvider? GetProvider(CompletionItem item)
+            public ImmutableArray<CompletionProvider> GetProviders(ImmutableHashSet<string> roles, Project? project)
+            {
+                using var _ = ArrayBuilder<CompletionProvider>.GetInstance(out var providers);
+                providers.AddRange(GetImportedAndBuiltInProviders(roles));
+                providers.AddRange(GetProjectCompletionProviders(project));
+                return providers.ToImmutable();
```

Comment:
❔ How hot is this path? I worry about the allocation costs it may have for using `ArrayBuilder` instead of `ConcatFast` like the filtered providers below.

Reply:
This is only called under a lock when we have a cache miss in `_rolesToProviders`, so once per roles

Reply:
This is only used in CompletionService.TestAccessor